### ISSUE TITLE
DockerCreateContainer.exposedPorts should not be a Map

### ISSUE THREAD

**Issue by febelectron** [bmuschko/gradle-docker-plugin#128]
Currently DockerCreateContainer.exposedPorts is a Map<String, Integer>, where key is a protocol (tcp/udp) and a value is a port number. This makes impossible to specify more than one same-protocol ports, e.g. if I'd like open web-port and JMX port at the same time.

https://docs.docker.com/engine/reference/api/docker_remote_api_v1.20/ says:
> ExposedPorts - An object mapping ports to an empty object in the form of: 
> "ExposedPorts": { "port/tcp|udp: {}" }

**Comment by bmuschko** [comment]
OK, thanks for the info. I wasn't aware of that. Would you be interested in providing a pull request. We could represent `ExposedPorts` as inner class of the task.

**Comment by cdancy** [comment]
@bmuschko to keep things ... simple ... on the task we could define exposedPorts to be something like:         

    Map<String, ?> exposedPorts = [:].withDefault{key -> return []}

Trying to keep things as groovy as possible with the above. We could make it more java-esque but then the initialization of the property would start to look messy IMO. The caller would then be able to do the following: 

    exposedPorts = ['tcp' : [123, 456, 789], 'udp' : [987, 654, 321]]

This would go against the standard set by the plugin and its tasks in that: if a property is set use it otherwise ignore. In this case we could make an exception and also check if the property is set and has size greater than or equal to 1. Documentation/explanation would be required to explain what we are doing but that should be fine for folks using this feature?

If we are using an inner class I could see, at least for this use case, things starting to get messy quick and not as simple as the above.

@febelectron @bmuschko thoughts one way or the other?

**Comment by cdancy** [comment]
On further testing this implementation would allow for the caller to pass in anything. Maybe an inner class would be best but coded in such way that we could do the following:

    exposedPorts = ExposedPorts.newInstance().expose('tcp', [123, 456,789]).expose('udp', [987,654,321])

This feels like a better solution and one we could guard against code-wise in the plugin.

**Comment by bmuschko** [comment]
I'd expose the property but also a method that makes it convenient to add new instances.

    class DockerCreateContainer {
        @Nested
        List<ExposedPort> exposedPorts = []

        void exposePort(String protocol, List<Integer> ports) {
            exposedPorts << new ExposedPort(protocol, ports)
        }

        static class ExposedPort {
            @Input
            String protocol

            @Input
            List<Integer> ports     

            ExposedPort(String protocol, List<Integer> ports) {
                this.protocol = protocol
                this.ports = ports
            }      
        }
    }

**Comment by cdancy** [comment]
I like it. Any objections to changing this:
```
    void exposePort(String protocol, List<Integer> ports) {
        exposedPorts << new ExposedPort(protocol, ports)
    }
```
to this:
```
    def exposePort(String protocol, List<Integer> ports) {
        exposedPorts << new ExposedPort(protocol, ports)
        this
    }
```
I'm a fan of the builder design pattern. It would allow folks to have the option to chain calls instead of listing them.

I can also send in the PR for this. It technically is a bug and so something we should fix if we've got the spare cycles. I've done work in this area recently so this should be no big deal getting patched up and writing a quick test for.

**Comment by bmuschko** [comment]
Configuring tasks with the builder pattern isn't a common practice. I'd stay away from it. In Groovy is pretty easy to achieve pretty much the same thing with the `with` statement. The following example configures a task named `createContainer`.

    createContainer.with {
        exposePort 'tcp', [123, 456, 789]
        exposePort 'udp', [987,654,321]
    }

**Comment by cdancy** [comment]
Fair enough. Let me put together a PR to wrap this up.

**Comment by cdancy** [comment]
Closing as issue has been addressed with #140
